### ISSUE TITLE
Can I add my own learning rate scheduler during training? 

### ISSUE THREAD

**Issue by weikunhan** [KevinMusgrave/powerful-benchmarker#80]
Hello Kevin @KevinMusgrave , 

1st question:

May I ask it is possible to add my own learning rate scheduler during training? It looks like you are using a constant learning rate. 

2nd question:

To fine-tune the pre-trained network on my own dataset, do we have to use a constant learning rate for Metric Learning? You mentioned in your paper: "
All network parameters are optimized using RMSprop with learning rate
1e-6. We chose RMSprop because it converges faster than SGD, and seems
to generalize better than Adam, based on a small set of experiments. For
loss functions that include their own learnable weights (e.g. ArcFace), we use
RMSprop but leave the learning rate as a hyperparameter to be optimized."

If using a constant learning rate, could you give me a reason why doing that?

Thank you 

Mark

**Comment by KevinMusgrave** [comment]
### 1. Register the factory

This is a bug (issue #72) I haven't gotten around to fixing. You can get lr schedulers working by replacing the last 3 lines of run.py with this:

```python
from powerful_benchmarker.factories import OptimizerFactory

class OptimizerWithSchedulerFactory(OptimizerFactory):
    def format_named_specs_output(self, named_specs_output):
        optimizers, lr_schedulers, gradient_clippers = {}, {}, {}
        for k1,v1 in named_specs_output.items():
            basename = k1.replace("_optimizer", '')
            if "optimizer" in v1: optimizers[k1] = v1["optimizer"]
            if "lr_scheduler" in v1:
                for k2, v2 in v1["lr_scheduler"].items():
                    lr_schedulers[k2] = v2
            if "gradient_clipper" in v1: gradient_clippers["{}_grad_clipper".format(basename)] = v1["gradient_clipper"]
        return optimizers, lr_schedulers, gradient_clippers


r = runner(**(args.__dict__))
r.register("factory", OptimizerWithSchedulerFactory)
r.run()
```

### 2. Use the factory

Then use this flag at the command line when you call run.py:
```
--factories {optimizer~OVERRIDE~: {OptimizerWithSchedulerFactory: {}}}
```

Or if you have your own directory of yaml files and have set ```--root_config_folder```, then you can change the [```default.yaml``` in ```factories```](https://github.com/KevinMusgrave/powerful-benchmarker/blob/master/src/powerful_benchmarker/configs/config_factories/default.yaml):
```yaml
  optimizer:
    OptimizerWithSchedulerFactory: {}
```

### 3. Use an lr scheduler
Schedulers are specified as an argument of the corresponding optimizer. For example, here's an exponential decay scheduler for the trunk optimizer that is called at every iteration:
```
--optimizers {trunk_optimizer: {scheduler_by_iteration: {ExponentialLR: {gamma: 0.99}}}}
```

Or in yaml form:
```yaml
optimizers:
  trunk_optimizer:
    RMSprop:
      lr: 0.000001
      weight_decay: 0.0001
      momentum: 0.9
    scheduler_by_iteration:
      ExponentialLR:
        gamma: 0.99
```

The 3 scheduler options are:
- ```scheduler_by_iteration```, stepped at every iteration
- ```scheduler_by_epoch```, stepped at every epoch
- ```scheduler_by_plateau```, stepped when the primary accuracy metric plateaus

You can use all 3 at once if you want to.

### 4. Why did I use a constant learning rate in the paper?
I wanted to keep the approach as simple as possible, so I used a constant low learning rate. There are many possible lr scheduling strategies, and I didn't want to add that to the hyperparameter search space. Also, I believe most metric learning papers use a constant learning rate.

That said, I don't think there is any theoretical reason why you can't use an lr scheduler when fine-tuning. So it's reasonable to try it out.

**Comment by weikunhan** [comment]
@KevinMusgrave Thank you so much for such a detailed replay. I appreciate that! I currently use some of your open source code in my research work. Your open-source project really helped me!

Mark

**Comment by KevinMusgrave** [comment]
Glad to hear that!
